src/frequenz/sdk/timeseries/formulas/_base_ast_node.py

```diff
@@ -0,0 +1,135 @@
+# License: MIT
+# Copyright © 2025 Frequenz Energy-as-a-Service GmbH
+
+"""Formula AST node base class."""
+
+import abc
+import asyncio
+import logging
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Generic
+
+from typing_extensions import override
+
+from ...timeseries import Sample
+from ...timeseries._base_types import QuantityT
+
+_logger = logging.getLogger(__name__)
+
+
+@dataclass(kw_only=True)
+class AstNode(abc.ABC, Generic[QuantityT]):
+    """An abstract syntax tree node representing a formula expression."""
+
+    span: tuple[int, int] | None = None
+    """The span (start, end) of the expression in the input string."""
+
+    @abc.abstractmethod
+    async def evaluate(self) -> Sample[QuantityT] | QuantityT | None:
+        """Evaluate the expression and return its numerical value."""
+
+    @abc.abstractmethod
+    def format(self, wrap: bool = False) -> str:
+        """Return a string representation of the node."""
+
+    @override
+    def __str__(self) -> str:
+        """Return the string representation of the node."""
+        return self.format()
+
+    @abc.abstractmethod
+    async def subscribe(self) -> None:
+        """Subscribe to any data streams needed by this node."""
+
+    @abc.abstractmethod
+    async def unsubscribe(self) -> None:
+        """Unsubscribe from any data streams used by this node."""
+
+
+class NodeSynchronizer(Generic[QuantityT]):
```

Comment:
It looks like you are always using the `NodeSynchronizer` once, and I think that makes sense. Any reason not to make it just a function, like `synchronize_evaluation(nodes: list[AstNode[QuantityT]], sync_to_first_node: bool = False)`?